docs/architecture/0e-plans.md

```diff
@@ -0,0 +1,474 @@
+
+At this point, we do not consider this critical for TF-PSA-Crypto 1.0, and we are likely to work on it during the lifetime of TF-PSA-Crypto 1.x. This may change if we reevaluate the priority based on concrete use cases. In the meantime, `oid.h` will become private, but will remain used in Mbed TLS.
+
+#### Platform headers
+
+Due to a lack of bandwidth, we are not planning any major changes to most platform interfaces. As of the 0ε target, the following platform-related headers remain public.
+
+* `memory_buffer_alloc.h`: a feature used by some applications, with close to 0 maintenance cost for us.
+* `platform.h`, `platform_time.h`, `platform_util.h`: platform abstractions. We would like to adapt them, but we do not think we will have time in before 1.0, so this will have to wait until the next major version.
+* `threading.h`: We want to change the threading abstraction, but this is out of scope of the 0ε project. https://github.com/Mbed-TLS/mbedtls/issues/8455
+
+### Plans for hiding
+
+In an exposed header, the minimum work for 0ε is:
+
+* Ensure that exposed interfaces are not listed in the rendered Doxygen documentation.
+* Ensure that applications cannot inadvertently call private functions that are declared in exposed headers.
+
+#### TF-PSA-Crypto header locations
+
+We distinguish between three categories of headers:
+
+* Public headers define public, stable APIs.
+* Exposed headers define exposed interfaces, as well as private interfaces used by Mbed TLS. They need to be present when building Mbed TLS.
+* Private headers are only needed when building TF-PSA-Crypto itself.
+
+The following table summarizes the characteristics of each category.
+
+| Category | Location | In include path? |
+| -------- | -------- | ---------------- |
+| Public   | `include` | yes |
+| Exposed  | `include` | yes except for Doxygen |
+| Private  | `drivers/*/include` | only for crypto |
+
+TODO: Task to set up the Doxygen build
+
+#### Hiding functions in an exposed header
+
+For TF-PSA-Crypto 0ε, all private functions defined in private headers will be guarded by `defined(MBEDTLS_DECLARE_PRIVATE_IDENTIFIERS)`. We define this macro when building TF-PSA-Crypto and Mbed TLS, but applications should not define it.
+
+This is similar to `MBEDTLS_ALLOW_PRIVATE_ACCESS` to “bless” access to structure fields.
+
+Prototype: https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/132
+
+Thus, for the 0ε milestone, in each affected header, we just need to place `#ifdef` guards around the function declarations.
+
+Rationale:
+
+* This is a small amount of work.
+* This requires few changes to the code, and they are very localized, so it will not disrupt other work happening in parallel.
+
+TODO: this leaves Doxygen comments around
```

Comment:
I think Doxygen obeys `#ifdef` (otherwise we wouldn't bother with `config.py realfull` in `apidoc_full.sh`), so can we find a way to ensure `MBEDTLS_DECLARE_PRIVATE_IDENTIFIERS` is not defined when we run doxygen? Like, in `common.h`:
```c
#if !defined(__DOXYGEN__)
#define MBEDTLS_DECLARE_PRIVATE_IDENTIFIERS
#endif
```
(Note: `__DOXYGEN__` is defined by `doxygen/mbedtls.doxyfile`.)

Wdyt?

Reply:
I meant, but forgot to write: Doxygen comments in the rest of the file, e.g. the `\file` comment and comments on the exposed types.

Reply:
I don't think we'd need a `!__DOXYGEN__` guard: we wouldn't be feeding `#define MBEDTLS_DECLARE_PRIVATE_IDENTIFIERS` to Doxygen since that would only appear when building library files.